doc/adr/0004-pab-indexing-solution-integration.rst

```diff
@@ -0,0 +1,183 @@
+The indexed data can be queried using the REST API interface.
+
+This brings up the main issue: the PAB and plutus-chain-index each listen to the same source of information (a local Cardano node), but each index the information at different speeds.
+For a dApp developer writing off-chain code using the Contract API, there is no abstraction for handling multiple sources of truth.
+
+Currently, in the best case scenario (fully synced PAB and plutus-chain-index), plutus-chain-index will always trail behind the in-memory storage of the PAB by a few seconds.
+Therefore, even in this scenario, querying the plutus-chain-index with `unspentTxOutFromRef` in the above contract has a high probability of returning `Nothing`.
+
+Decisions
+---------
+
+The best solution is probably a combination of the `Alternative solutions`_ described below.
+However, we will mainly choose the `Query functions should interact with a single source of truth`_ solution.
+
+* We will replace `plutus-chain-index` with `Marconi` as PAB's indexing component
+
+* We will move out the Blockchain information indexed by PAB in Marconi
+
+* We will add new indexers in Marconi in order to replicate the information indexed by `plutus-chain-index`
+
+* We will adapt the architecture of Marconi (which will become our new indexing component) to support waiting queries
+
+* Since HTTP APIs are not well suited to waiting queries, we will use RPC via UNIX Socket to communicate information between PAB and Marconi
+
+Alternative solutions
+---------------------
+
+In this section, we describe all the ways to deal with the problem.
+Note that final decision might include one or more of these alternate solutions.
+
+Make sure all components are in sync
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We can make sure that all indexing components are syncing at the same speed.
+For example, if PAB syncs from the local node and arrives at slot 100, then it needs to wait for the chain-index to also arrive at slot 100.
+Only then can it respond to a Contract request.
+
+A simply way to achieve this behavior is to change the implementation of Contract API handler functions (like `utxosAt`, `unspentTxOutFromRef`, etc.) by waiting for the component to be in sync with all other components.
+For example, let's take the `utxosAt` request which basically queries UTXOs from a given address using `plutus-chain-index`.
+We could, before querying, wait for `plutus-chain-index` to be in sync with all components which index blockchain data (PAB, Blockfrost_).
+
+Pros:
+
+- All the indexed information is consistent with each other
+
+Cons:
+
+- As fast as the slowest indexing component
+- Tight coupling between the indexing components meaning that if the Contract only uses chain-index requests without using requests from other indexing components, the chain-index will still have to wait for all other components to be in sync with each other
```

Comment:
maybe the interpreter can be given the list of relevant indexers?

Reply:
Problem with that is that the PAB doesn't know what the "relevant" indexers are (unless they are specified by the user when starting the PAB). However, doing it that way will make the PAB highly coupled with the builtin contracts. Might not be such a problem right now, but imagine we extend the PAB to be even more generalized, where the contracts are not builtin, but are actually separate executable files which can be injected in PAB.

Therefore, I believe PAB should not care about what requests the contract is composed of.